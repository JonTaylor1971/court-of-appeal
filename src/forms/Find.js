import React, { Component } from 'react';
import DefaultService from '../service/default.service.js';
import NumericField from '../components/fields/NumericField';
import SpinnerButton from '../components/SpinnerButton';

class Find extends Component {

    constructor(props) {
        super(props);
        this.service = props.service;
        this.callback = props.callback;
        this.search = this.search.bind(this);
        this.handleFieldChange = this.handleFieldChange.bind(this);
        this.isInvalidCaseNumber = this.isInvalidCaseNumber.bind(this);
        this.state = {
            searchDisabled: true
        };
    }

    componentDidMount() {
        if (this.service == null) {
            let window = this.element.ownerDocument.defaultView;
            this.service = new DefaultService(window);
        }
    }

    search() {
        this.props.startSearching();
        this.printButton.startSpinner();
        this.service.searchForm7(this.props.formSevenNumber, (data) => {
            this.printButton.stopSpinner();
            this.callback(data);
        });
    }

    handleKeyPress(target) {
        if(target.charCode === 13 && !this.isInvalidCaseNumber(this.props.formSevenNumber)){
            this.search();
        }
    }

    render() {
        return (
            <div className="form-section not-printable" ref={ (element)=> {this.element = element }}>
                <h2 style={{ fontWeight:'bold' }}>Appeal Case Information</h2>
                Enter the Court of Appeal number you want to respond to:

                <table>
                    <tbody>
                        <tr>
                            <td>LEVEL OF COURT</td>
                            <td>COURT OF APPEAL FILE NO.</td>
                            <td> </td>
                        </tr>
                        <tr>
                            <td><input disabled value="Court of Appeal" /></td>
                            <td>
                                <NumericField
                                    id="file-no"
                                    name="number.form-seven"
                                    autofocus={true}
                                    value={this.props.formSevenNumber}
                                    handleFieldChange={this.handleFieldChange}
                                    handleKeyPress={this.handleKeyPress.bind(this)}
                                />
                            </td>
                            <td>
                            <SpinnerButton disabled={this.state.searchDisabled} id="find-button" width="52" onClick={this.search} ref={ (element)=> {this.printButton = element }}
                                content='Find'>
                            </SpinnerButton>
                            </td>
                            <td>
                                <div className="error-message">
                                {this.props.notFoundError}
                                </div>
                            </td>
                        </tr>
                    </tbody>
                </table>
            </div>
        );
    }

    handleFieldChange(e) {
        // If the first two digits are not CA, add a CA.  otherwise only accept ^CA\d{5}\d*sea$
        e.target.value = 'CA'.concat(e.target.value.replace(/\D/g,''));
        if (this.isInvalidCaseNumber(e.target.value)) {
            this.setState({ searchDisabled: true });
        } else {
            this.setState({ searchDisabled: false });
        }
        this.props.handleFieldChange(e);
    }

    isInvalidCaseNumber(caseNumber) {
        return caseNumber.match(/^CA\d{5}\d*$/) === null;
    }
}

export default Find;
