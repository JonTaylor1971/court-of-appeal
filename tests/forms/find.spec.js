require('../support/enzyme.setup');
import React from 'react';
import Find from '../../src/forms/Find';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

test('valid entries', ()=> {

    let value = '', started = false;
    let handleFieldChange = (e) => { value = e.target.value; };
    let startSearch = () => { started = true };

    const finder = renderer.create(
        <Find id="number-field"
              formSevenNumber={value}
              service={{ searchForm7: ()=> {} }}
              handleFieldChange={handleFieldChange.bind(this)}
              startSearching={startSearch.bind(this)}
        />,
    );
    let tree = finder.toJSON();
    expect(tree).toMatchSnapshot();

    let instance = finder.getInstance();

    let e = { target: { value: "CA12345"}};
    instance.handleFieldChange(e);
    expect(value).toMatch('CA12345');

    e = { target: { value: "CAT IN HAT2" } };
    instance.handleFieldChange(e);
    expect(value).toMatch('CA2');

    e = { target: { value: '1234' } };
    instance.handleFieldChange(e);
    expect(value).toMatch('CA1234');

    e = { target: { value: "*bah1#?234?wtf?  5" } };
    instance.handleFieldChange(e);
    expect(value).toMatch('CA12345');

});

test('cannot search with invalid entries', ()=> {

    let value = '', started = false;
    let handleFieldChange = (e) => { value = e.target.value; };
    let startSearch = () => { started = true };

    const finder = renderer.create(
        <Find id="number-field"
              formSevenNumber={value}
              service={{ searchForm7: ()=> {} }}
              handleFieldChange={handleFieldChange.bind(this)}
              startSearching={startSearch.bind(this)}
        />,
    );
    let tree = finder.toJSON();
    expect(tree).toMatchSnapshot();

    let instance = finder.getInstance();

    let e = { target: { value: "CA12345"}};
    instance.handleFieldChange(e);
    expect(value).toMatch('CA12345');
    expect(instance.state.searchDisabled).toEqual(false);

    e = { target: { value: "CA1" } };
    instance.handleFieldChange(e);
    expect(value).toMatch('CA1');
    expect(instance.state.searchDisabled).toEqual(true);

});


let DefaultService = require('../../src/service/default.service');
test('default service', ()=>{
    let value = '', started = false;
    let handleFieldChange = (e) => { value = e.target.value; };
    let startSearch = () => { started = true };
    const finder = mount(
        <Find id="number-field"
              formSevenNumber={value}
              handleFieldChange={handleFieldChange.bind(this)}
              startSearching={startSearch.bind(this)}
        />
    );    
    let instance = finder.instance();
    
    expect(instance.service instanceof DefaultService).toEqual(true);  
});
test('sends fetched data to caller', ()=>{
    let value = '', started = false;
    let handleFieldChange = (e) => { value = e.target.value; };
    let startSearch = () => { started = true }
    let sent = {};
    let callback = (data) => { sent=data; }
    const finder = renderer.create(
        <Find id="number-field"
              formSevenNumber={value}
              service={{ searchForm7: (file, callback)=> {callback(42); } }}
              callback={callback.bind(this)}
              handleFieldChange={handleFieldChange.bind(this)}
              startSearching={startSearch.bind(this)}
        />,
    );    
    let instance = finder.getInstance();
    instance.search();

    expect(sent).toEqual(42);
});
test('[enter] can trigger the search', ()=>{
    let value = 'CA12345', started = false;
    let handleFieldChange = (e) => { value = e.target.value; };
    let startSearch = () => { started = true };
    let sent = {};
    let callback = (data) => { sent=data; }
    const finder = renderer.create(
        <Find id="number-field"
              formSevenNumber={value}
              service={{ searchForm7: (file, callback)=> {callback(42); } }}
              callback={callback.bind(this)}
              handleFieldChange={handleFieldChange.bind(this)}
              startSearching={startSearch.bind(this)}
        />,
    );    
    let instance = finder.getInstance();
    instance.handleKeyPress({charCode:13});

    expect(sent).toEqual(42);
});
test('only [enter] can trigger the search', ()=>{
    let value = 'CA12345', started = false;
    let handleFieldChange = (e) => { value = e.target.value; };
    let startSearch = () => { started = true }
    let sent = {};
    let callback = (data) => { sent=data; };
    const finder = renderer.create(
        <Find id="number-field"
              formSevenNumber={value}
              service={{ searchForm7: (file, callback)=> {callback(42); } }}
              callback={callback.bind(this)}
              handleFieldChange={handleFieldChange.bind(this)}
              startSearching={startSearch.bind(this)}
        />,
    );
    let instance = finder.getInstance();
    instance.handleKeyPress({charCode:65});

    expect(sent).toEqual({});
});
