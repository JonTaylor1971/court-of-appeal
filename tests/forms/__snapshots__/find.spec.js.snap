// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valid entries 1`] = `
<div
  className="form-section"
>
  <h2
    style={
      Object {
        "fontWeight": "bold",
      }
    }
  >
    Appeal Case Information
  </h2>
  Enter the Court of Appeal number you want to respond to:
  <table>
    <tbody>
      <tr>
        <td>
          LEVEL OF COURT
        </td>
        <td>
          COURT OF APPEAL FILE NO.
        </td>
        <td>
           
        </td>
      </tr>
      <tr>
        <td>
          <input
            disabled={true}
            value="Court of Appeal"
          />
        </td>
        <td>
          <input
            autoFocus={true}
            id="file-no"
            name="number.form-seven"
            onChange={[Function]}
            onKeyPress={[Function]}
            placeholder={undefined}
            readOnly={undefined}
            value=""
          />
        </td>
        <td>
          <button
            className="btn btn-primary btn-green"
            disabled={true}
            id="find-button"
            onClick={[Function]}
          >
            Find
          </button>
        </td>
        <td>
          Error message will display here when we hook it up.
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
